server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -207,30 +208,31 @@ private static boolean isEmptyDoc(RootObjectMapper root, XContentParser parser)
                 case FIELD_NAME:
                     return false;
                 default:
-                    throwOnMalformedContent();
+                    throwOnMalformedContent(parser);
             }
         }
         return false;
     }
 
-    private static void throwOnMalformedContent() {
-        throw new MapperParsingException(
+    private static void throwOnMalformedContent(XContentParser parser) {
+        throw new DocumentParsingException(
+            parser.getTokenLocation(),
             "Malformed content, after first object, either the type field or the actual properties should exist"
         );
     }
 
-    private static MapperParsingException wrapInMapperParsingException(SourceToParse source, Exception e) {
+    private static DocumentParsingException wrapInDocumentParsingException(SourceToParse source, Exception e) {
         // if its already a mapper parsing exception, no need to wrap it...
-        if (e instanceof MapperParsingException) {
-            return (MapperParsingException) e;
+        if (e instanceof DocumentParsingException) {
+            return (DocumentParsingException) e;
         }
 
         // Throw a more meaningful message if the document is empty.
         if (source.source() != null && source.source().length() == 0) {
-            return new MapperParsingException("failed to parse, document is empty");
+            return new DocumentParsingException(XContentLocation.UNKNOWN, "failed to parse, document is empty");
         }
 
-        return new MapperParsingException("failed to parse", e);
+        return new DocumentParsingException(XContentLocation.UNKNOWN, "failed to parse: " + e.getMessage(), e);
```

Comment:
We can't really provide a more useful location in these cases?

Reply:
In some cases there's no information, but I've reworked things a bit so that we can catch jackson errors that do include location data and include that in the error message.